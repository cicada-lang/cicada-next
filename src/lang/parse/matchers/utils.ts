import * as Exps from "../../exp"

export function piBindingtoFnBindingFrom(binding: Exps.PiBinding): Exps.FnBinding {
  switch (binding.kind) {
    case "PiBindingNameless": {
      return Exps.FnBindingName("_")
    }

    case "PiBindingNamed": {
      return Exps.FnBindingName(binding.name)
    }

    case "PiBindingImplicit": {
      return Exps.FnBindingImplicit(binding.name)
    }
  }
}
