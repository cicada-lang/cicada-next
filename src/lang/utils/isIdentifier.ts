const identifierRegExp = new RegExp("^([-_\\p{Letter}][-_\\p{Letter}0-9]*)$", "u")

export function isIdentifier(text: string): boolean {
  return identifierRegExp.test(text)
}
