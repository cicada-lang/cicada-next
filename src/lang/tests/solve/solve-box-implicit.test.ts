import { expect, test } from "vitest"
import { runCode } from "../utils"

test("solve Box -- implicit", async () => {
  const output = await runCode(`

function Box(T: Type): Type {
  return Pair(T, Trivial)
}

function box(implicit T: Type, x: T): Box(T) {
  return cons(x, sole)
}

solve (x: String, y: Box(String), z: Box(Box(String))) {
  unify y = box(x)
  unify z = box(y)
  unify x = "abc"
}

`)

  expect(output).toMatchInlineSnapshot(
    '"{ x: \\"abc\\", y: cons(x, sole), z: cons(cons(x, sole), sole) }"',
  )
})
