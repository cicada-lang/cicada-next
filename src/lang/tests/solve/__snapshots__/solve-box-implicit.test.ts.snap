// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`solve Box -- implicit > 
    "{
      x: \\"abc\\",
      y: cons(\\"abc\\", sole),
      z: cons(cons(\\"abc\\", sole), sole)
    }"
   1`] = `
"{
  x: \\"abc\\",
  y: cons(\\"abc\\", sole),
  z: cons(cons(\\"abc\\", sole), sole)
}"
`;

exports[`solve Box -- implicit: 
    "{
      x: "abc",
      y: cons("abc", sole),
      z: cons(cons("abc", sole), sole)
    }"
   1`] = `
"{
  x: "abc",
  y: cons("abc", sole),
  z: cons(cons("abc", sole), sole)
}"
`;

exports[`solve Box -- implicit 1`] = `
"{
  x: "abc",
  y: cons("abc", sole),
  z: cons(cons("abc", sole), sole)
}"
`;
